### PR TITLE
Add the ability to set the passphrase secrets provider to read from a passphrase file

### PR DESCRIPTION
**Pull request by stack72**
Fixes: #5305

https://asciinema.org/a/INd7fs6nALeuKgxusIxr73VHk

### CODE REVIEW

File: pkg/cmd/pulumi/crypto_local.go [hunk 1]
```diff
@@ -35,8 +37,20 @@ func readPassphrase(prompt string) (string, error) {
 	if phrase, ok := os.LookupEnv("PULUMI_CONFIG_PASSPHRASE"); ok {
 		return phrase, nil
 	}
+	if phraseFile, ok := os.LookupEnv("PULUMI_CONFIG_PASSPHRASE_FILE"); ok {
+		phraseFilePath, err := filepath.Abs(phraseFile)
+		if err != nil {
+			return "", errors.New("unable to construct a path the PULUMI_CONFIG_PASSPHRASE_FILE")
+		}
+		phraseDetails, err := ioutil.ReadFile(phraseFilePath)
```
Comment (justinvp) [inline pkg/cmd/pulumi/crypto_local.go:45]: What if the file has trailing newlines? I assume we do not want to treat that as part of the passphrase.

Maybe do it like how we read from the console elsewhere (which can be from a file via redirected std input)?

https://github.com/pulumi/pulumi/blob/848a98d5112ec93239501fd66b51c9f1f30353ac/sdk/go/common/util/cmdutil/console.go#L72-L78

**Review by justinvp — APPROVED** [review]
LGTM otherwise.

Also, mind opening a PR in the docs repo to add the new environment variable to the docs? https://github.com/pulumi/docs/blob/master/content/docs/reference/cli/environment-variables.md

File: pkg/cmd/pulumi/crypto_local.go [hunk 2]
```diff
@@ -35,8 +37,27 @@ func readPassphrase(prompt string) (string, error) {
...
+			return "", errors.New("unable to open PULUMI_CONFIG_PASSPHRASE_FILE")
+		}
+
+		reader := bufio.NewReader(f)
+		passphraseDetails, err := reader.ReadString('\n')
+		if err != nil {
+			return "", errors.New("unable to read PULUMI_CONFIG_PASSPHRASE_FILE")
+		}
+
+		return cmdutil.RemoveTrailingNewline(passphraseDetails), nil
```
Comment (justinvp) [inline pkg/cmd/pulumi/crypto_local.go:56]: Sorry, looking this over now, prob easiest to read it as before, but just trim the string before returning it as Lee suggested.

File: pkg/cmd/pulumi/crypto_local.go [hunk 3]
```diff
@@ -35,8 +37,27 @@ func readPassphrase(prompt string) (string, error) {
 	if phrase, ok := os.LookupEnv("PULUMI_CONFIG_PASSPHRASE"); ok {
 		return phrase, nil
 	}
+	if phraseFile, ok := os.LookupEnv("PULUMI_CONFIG_PASSPHRASE_FILE"); ok {
+		phraseFilePath, err := filepath.Abs(phraseFile)
+		if err != nil {
+			return "", errors.New("unable to construct a path the PULUMI_CONFIG_PASSPHRASE_FILE")
```
Comment (justinvp) [inline pkg/cmd/pulumi/crypto_local.go:43]:
```suggestion
			return "", errors.Wrap(err, "unable to construct a path the PULUMI_CONFIG_PASSPHRASE_FILE")
```

(Suggestion applies to all errors)

Comment (justinvp) [reply pkg/cmd/pulumi/crypto_local.go:44]: This requires changing the import to `"github.com/pkg/errors"`

File: pkg/cmd/pulumi/crypto_local.go (hunk 2 above)
Comment (stack72) [reply pkg/cmd/pulumi/crypto_local.go]: fixed!

**Pull request merged**